firedrake/adjoint/blocks.py

```diff
@@ -587,3 +587,144 @@ def prepare_recompute_component(self, inputs, relevant_outputs):
 
     def recompute_component(self, inputs, block_variable, idx, prepared):
         return self.backend.interpolate(prepared, self.V)
+
+
+class SupermeshProjectBlock(Block, Backend):
+    r"""
+    Annotates supermesh projection.
+
+    Suppose we have a source space, :math:`V_A`, and a target space, :math:`V_B`.
+    Projecting a source from :math:`V_A` to :math:`V_B` amounts to solving the
+    linear system
+
+ .. math::
+        M_B * v_B = M_{AB} * v_A,
+
+    where :math:`M_B` is the mass matrix on :math:`V_B`, :math:`M_{AB}` is the
+    mixed mass matrix for :math:`V_A` and :math:`V_B` and :math:`v_A` and
+    :math:`v_B` are vector representations of the source and target
+    :class:`.Function`s.
+
+    This can be broken into two steps:
+      Step 1. form RHS, multiplying the source with the mixed mass matrix;
+      Step 2. solve linear system.
+    """
+    def __init__(self, source, target_space, target, bcs=[], *args, **kwargs):
+        super(SupermeshProjectBlock, self).__init__()
+        import firedrake.supermeshing as supermesh
+        from firedrake.utils import complex_mode, SLATE_SUPPORTS_COMPLEX
+
+        # Process args and kwargs
+        if not isinstance(source, self.backend.Function):
+            raise NotImplementedError(f"Source function must be a Function, not {type(source)}.")
+        sp = kwargs.get('solver_parameters')
+        self.solver_parameters = {} if sp is None else sp.copy()
+        self.form_compiler_parameters = kwargs.get('form_compiler_parameters')
+        self.constant_jacobian = kwargs.get('constant_jacobian', True)
+        use_slate_for_inverse = kwargs.get('use_slate_for_inverse', True)
+        try:
+            element = target_space.finat_element
+            is_dg = element.entity_dofs() == element.entity_closure_dofs()
+            is_variable_layers = target_space.mesh().variable_layers
+        except AttributeError:
+            is_dg = False
+            is_variable_layers = True
+        self.use_slate_for_inverse = (use_slate_for_inverse and is_dg and not is_variable_layers
+                                      and (not complex_mode or SLATE_SUPPORTS_COMPLEX))
+
+        # Store spaces
+        mesh = kwargs.pop("mesh", None)
+        if mesh is None:
+            mesh = target_space.mesh()
+        self.source_space = source.function_space()
+        self.target_space = target_space
+
+        # Form LHS
+        dx = self.backend.dx(domain=mesh)
+        w = self.backend.TestFunction(target_space)
+        Pv = self.backend.TrialFunction(target_space)
+        self.lhs = self.backend.assemble(
+            self.backend.inner(Pv, w)*dx, bcs=bcs,
+            mat_type=self.solver_parameters.get("mat_type"),
+            form_compiler_parameters=self.form_compiler_parameters)
```

Comment:
This seems to just duplicate code from a projector. Why not make a projector?

Reply:
Concretely, make a `self.projector = Projector(...)` and use self.projector.apply_massinv instead of duplicating code.

Reply:
Thanks, will do